packages/grafana-data/src/transformations/transformers/filterByValue.ts

```diff
@@ -0,0 +1,133 @@
+import { map } from 'rxjs/operators';
+
+import { noopTransformer } from './noop';
+import { DataTransformerID } from './ids';
+import { DataTransformerInfo, MatcherConfig } from '../../types/transformations';
+import { DataFrame, Field } from '../../types/dataFrame';
+import { getFieldDisplayName } from '../../field/fieldState';
+import { getValueMatcher } from '../matchers';
+import { ArrayVector } from '../../vector/ArrayVector';
+
+export enum FilterByValueType {
+  exclude = 'exclude',
+  include = 'include',
+}
+
+export enum FilterByValueMatch {
+  all = 'all',
+  any = 'any',
+}
+
+export interface FilterByValueFilter {
+  fieldName: string;
+  config: MatcherConfig;
+}
+
+export interface FilterByValueTransformerOptions {
+  filters: FilterByValueFilter[];
+  type: FilterByValueType;
+  match: FilterByValueMatch;
+}
+
+export const filterByValueTransformer: DataTransformerInfo<FilterByValueTransformerOptions> = {
+  id: DataTransformerID.filterByValue,
+  name: 'Filter data by values',
+  description: 'select a subset of results based on values',
+  defaultOptions: {
+    filters: [],
+    type: FilterByValueType.include,
+    match: FilterByValueMatch.any,
+  },
+
+  operator: options => source => {
+    const filters = options.filters;
+    const matchAll = options.match === FilterByValueMatch.all;
+    const include = options.type === FilterByValueType.include;
+
+    if (!Array.isArray(filters) || filters.length === 0) {
+      return source.pipe(noopTransformer.operator({}));
+    }
+
+    return source.pipe(
+      map(data => {
+        if (!Array.isArray(data) || data.length === 0) {
+          return data;
+        }
+
+        const rows = new Set<number>();
+
+        for (const frame of data) {
+          const fieldIndexByName = groupFieldIndexByName(frame, data);
+
+          for (let index = 0; index < frame.length; index++) {
+            if (rows.has(index)) {
+              continue;
+            }
+
+            const checkIfFilterIsMatchingRow = (filter: FilterByValueFilter): boolean => {
+              const matcher = getValueMatcher(filter.config);
+              const fieldIndex = fieldIndexByName[filter.fieldName] ?? -1;
+              if (fieldIndex < 0) {
+                return false;
+              }
+              return matcher(index, frame.fields[fieldIndex], frame, data);
+            };
+
+            const matching = matchAll
+              ? filters.every(checkIfFilterIsMatchingRow)
+              : !!filters.find(checkIfFilterIsMatchingRow);
+
+            if (matching) {
+              rows.add(index);
+            }
+          }
+        }
```

Comment:
In the original implementation I was looping over the `filters` array first, and then on the rows. The reason was to avoid having to re-create the matcher function (given by `getValueMatcher()` here) for each row.

When you check each row for one filter at the time, you only have to call `getValueMatcher()` function once per filter (so `fiilter.length` times total). If you loop over the rows first like you are doing here, you have to call `filters.length` times the `getValueMatcher()` function **per row** (so `filter.length` x `rows.length` total), which seems much less optimal speed wise. Not sure how much of a difference it makes, but it certainly adds quite a few operation to execute inside the loop.

Reply:
Yeah, agree! I will look into making this a bit more performant.

Reply:
I have pushed some changes regarding this @torkelo @Totalus. Please have a look.